### PR TITLE
feat(client): add usage report config

### PR DESCRIPTION
**Pull request by fortuna**
This PR adds support for usage reporting. The config format looks like this:

```yaml
transport:
  $type: tcpudp
  tcp:
      $type: shadowsocks
      endpoint: example.com:80
      <<: &cipher
        cipher: chacha20-ietf-poly1305
        secret: SECRET
      prefix: "POST "
  udp:
      $type: shadowsocks
      endpoint: example.com:53
      <<: *cipher

reporter:
  $type: http
  url: https://your-callback-server.com/outline_callback
  interval: 1h
```

The `http` reporter will send a POST request to the given URL at the beginning of the session, and then every `interval`.

/cc @emohandesi @62w71st

### CODE REVIEW

File: client/go/outline/reporting/config.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+func NewHTTPReporterSubParser(httpClient *http.Client) func(ctx context.Context, input map[string]any) (Reporter, error) {
+	return func(ctx context.Context, input map[string]any) (Reporter, error) {
+		var config HTTPReporterConfig
+		if err := configyaml.MapToAny(input, &config); err != nil {
+			return nil, fmt.Errorf("invalid config format: %w", err)
+		}
+
+		collectorURL, err := url.Parse(config.URL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to report collector URL: %w", err)
```
Comment (jyyi1) [inline client/go/outline/reporting/config.go]: *nit*

```suggestion
			return nil, fmt.Errorf("failed to parse the report collector URL: %w", err)
```

File: client/go/outline/reporting/config.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		}
+		reporter := &HTTPReporter{URL: *collectorURL, HttpClient: httpClient}
+
+		if config.Interval != "" {
+			interval, err := time.ParseDuration(config.Interval)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse interval: %w", err)
+			}
+			if interval <= 0 {
+				return nil, fmt.Errorf("tunnel usage interval must be greater than 0")
```
Comment (jyyi1) [inline client/go/outline/reporting/config.go]: *nit*
```suggestion
				return nil, fmt.Errorf("http report interval must be greater than 0")
```

File: client/go/outline/reporting/config.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+			return nil, fmt.Errorf("invalid config format: %w", err)
+		}
+
+		collectorURL, err := url.Parse(config.URL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to report collector URL: %w", err)
+		}
+		reporter := &HTTPReporter{URL: *collectorURL, HttpClient: httpClient}
+
+		if config.Interval != "" {
```
Comment (jyyi1) [inline client/go/outline/reporting/config.go:46]: What would be the default `reporter.Interval` if no one is provided in the config? `0`? Will that cause any problems? [`time.NewTicker` might panic on `0`](https://cs.opensource.google/go/go/+/refs/tags/go1.25.0:src/time/tick.go;l=36)

File: client/go/outline/reporting/reporter.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			if !ok {
+				return
+			}
+			r.reportAndLogError()
+		}
+	}
+}
+
+func (r *HTTPReporter) reportAndLogError() {
+	slog.Debug("Sending report", "url", r.URL.String())
```
Comment (jyyi1) [inline client/go/outline/reporting/reporter.go]: *nit* I think `slog` could output `r.URL` directly, no need to convert to `String()`.

File: client/go/outline/client_test.go [hunk 5]
```diff
@@ -303,3 +305,32 @@ func Test_NewClientFromJSON_Errors(t *testing.T) {
...
+        secret: SECRET
+      prefix: "POST "
+  udp:
+      $type: shadowsocks
+      endpoint: example.com:53
+      <<: *cipher
+reporter:
+  $type: http
+  url: https://your-callback-server.com/outline_callback
+  interval: 10s`
```
Comment (jyyi1) [inline client/go/outline/client_test.go]: If `interval` is optional, we can add a test with that being empty.

File: client/go/outline/client.go [hunk 6]
```diff
@@ -124,5 +135,38 @@ func NewClientWithBaseDialers(clientConfigText string, tcpDialer transport.Strea
...
+				Cause:   platerrors.ToPlatformError(err),
+			}
+		}
+		client.reporter = reporter
+	}
+
+	return client, nil
+}
+
+func NewReporterParser(httpClient *http.Client) *configyaml.TypeParser[reporting.Reporter] {
```
Comment (jyyi1) [inline client/go/outline/client.go:165]: *nit* Shall we separate this out from `client.go`? And move it to another file?

**Review by jyyi1 — COMMENTED** [review]
Thanks for the implementation. Logic is clear, added a few comments.

File: client/go/outline/reporting/reporter.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+type HTTPReporter struct {
+	URL        url.URL
+	Interval   time.Duration
+	HttpClient *http.Client
+}
+
+func (r *HTTPReporter) Run(sessionCtx context.Context) {
+	r.reportAndLogError()
+	ticker := time.NewTicker(r.Interval)
```
Comment (fortuna) [inline client/go/outline/reporting/reporter.go:46]: To fix: only run the loop when the interval is set.
If interval is absent it should report only once

Comment (fortuna) [inline client/go/outline/reporting/reporter.go:46]: TODO: limit how frequent reporting happens to prevent DDoS.
We should also have an exponential backoff if the collector returns an error.
We can also consider some sort of authentication, where the client first verifies with the collector that it wants to receive reports (e.g. some sort of header or value in the response), before it sends any other report.

File: client/go/outline/reporting/config.go (hunk 3 above)
Comment (fortuna) [reply client/go/outline/reporting/config.go:46]: I'd like to make it report only once, on connection. I need to fix the reporter code for that.

File: client/go/outline/client.go (hunk 6 above)
Comment (fortuna) [reply client/go/outline/client.go:165]: Maybe. For now I think it's easier to keep it here, because it's going to change. We can move it later.

**Review by fortuna — COMMENTED** [review]
Thanks for the quick review. I'll try to address them tomorrow morning.

File: client/go/outline/reporting/reporter.go (hunk 7 above)
Comment (fortuna) [reply client/go/outline/reporting/reporter.go:46]: Foxed

Comment (fortuna) [reply client/go/outline/reporting/reporter.go:46]: Added 1h minimum. Backoff to be added

File: client/go/outline/client_test.go (hunk 5 above)
Comment (fortuna) [reply client/go/outline/client_test.go]: Done

File: client/go/outline/reporting/reporter.go (hunk 7 above)
Comment (jyyi1) [reply client/go/outline/reporting/reporter.go:46]: I think exponential backoff is optional if the minimum interval is `1h`?

**Review by jyyi1 — APPROVED** [review]
Thanks, LGTM now. It would be good to go after fixing the race condition caught by CI.

File: client/go/outline/reporting/reporter.go (hunk 7 above)
Comment (fortuna) [reply client/go/outline/reporting/reporter.go:46]: I agree. I'll punt on that for now. We can add it later.

**Comment by fortuna** [comment]
FYI, I had to fix an issue caused by the upgrade to 1.25 and Go Mobile: https://github.com/golang/go/issues/71827

**Comment by emohandesi** [comment]
What is the use case scenario for this reporter that sends a POST request to a server periodically?

**Comment by fortuna** [comment]
@emohandesi You guys requested it (every 24h). We can remove it though.

**Comment by emohandesi** [comment]
> @emohandesi You guys requested it (every 24h). We can remove it though.

OK, I see. I thought there was another use case for it. I was trying to figure out if we could somehow take advantage of it.

**Comment by sahin141** [comment]
The new update introduced usage reporting via the HTTP reporter, which sends a POST request to the configured URL at the start of the session and then at every interval. However, I noticed that it only sends the request if the connection is successful.

With this method, is it possible to also receive error messages or connection failure reports through the same reporting mechanism?

**Comment by emohandesi** [comment]
> The new update introduced usage reporting via the HTTP reporter, which sends a POST request to the configured URL at the start of the session and then at every interval. However, I noticed that it only sends the request if the connection is successful.
> 
> With this method, is it possible to also receive error messages or connection failure reports through the same reporting mechanism?

No, sending the connection error is not possible in this feature. #2036 and #2484 are implementing the error reporting feature.

**Comment by khantzawhein** [comment]
Hello, @fortuna, could you give me some insight into how this feature works and what the use case for it is?

I have been looking for this feature's documentation on Google Developer, but it seems like this feature has not been added to the documentation yet.

Edit: Doc found on https://github.com/Jigsaw-Code/outline-apps/blob/master/client/config.md
